aws/subscriber.go

```diff
@@ -278,38 +286,39 @@ func (s *awsSubscriber) ensureFilterPolicy(filter map[string]string) error {
 }
 
 // pull returns the message and error channel for the subscriber
-func (s *awsSubscriber) pull(ctx context.Context, channel chan pubsub.Message, errChannel chan error) {
-	s.wg.Add(1)
-	defer s.wg.Done()
+func (s *awsSubscriber) pull(ctx context.Context) ([]*awsMessage, error) {
 	resp, err := s.sqs.ReceiveMessageWithContext(ctx, &sqs.ReceiveMessageInput{
 		QueueUrl:              s.queueURL,
 		WaitTimeSeconds:       aws.Int64(20),
 		MaxNumberOfMessages:   aws.Int64(1),
 		MessageAttributeNames: []*string{aws.String("All")},
 	})
 	if err != nil {
-		errChannel <- err
-		return
+		log.Printf("AWS: error while fetching messages %v", err)
+		return nil, err
 	}
+
+	messages := make([]*awsMessage, 0, len(resp.Messages))
 	for _, msg := range resp.Messages {
 		message, err := decodeFromSQSMessage(msg.Body)
 		if err != nil {
 			log.Printf("AWS: error parsing SQS message body: %v", err)
-			errChannel <- err
-			continue
+			return nil, err
 		}
+
 		attributes, err := decodeMessageAttributes(msg.Body)
 		if err != nil {
 			log.Printf("AWS: error parsing SQS message attributes: %v", err)
-			errChannel <- err
-			continue
+			return nil, err
 		}
-		channel <- &awsMessage{
+		messages = append(messages, &awsMessage{
```

Comment:
Is there an advantage to not returning any of these messages if there are any errors, versus pushing the messages that don't have errors to the msgChannel before failing on a later error? It just seems a little odd to me to wrap the in a slice to return them, just to then iterator over the slicer to push them into the channel.